### DISCUSSION TITLE
Contributors wanted (not just code!), beginners welcome

### DISCUSSION THREAD

**Discussion by briantist** [hvac/hvac#994] [Announcements]
The `hvac` project has a **strong need** for help continuing to maintain the project. 
I'm looking to get more involvement from _anyone_ who has an interest in doing so.

# What does it mean to be a contributor?
First and foremost I want to stress that the **majority of the help needed is not coding!** There is plenty of coding to do but maintenance of the project is much more than that, and non-code contributions are essential.

## Hugely helpful things that you can do **today** to get involved
* [review PRs](https://github.com/hvac/hvac/pulls) -- this includes PRs from other contributors and from current maintainers like myself! yes I need reviews too, this is #1 on this list for a reason!
* [triage issues](https://github.com/hvac/hvac/issues) -- lots to do here even with varying levels of familiarity with the codebase
  * identifying whether something is a bug, where the bug might lie in the code, narrowing down the conditions to reproduce it, etc.
  * I cannot stress how useful **reproducing the issue yourself** can be, this can provide extremely valuable data points. For some types of issues, for example ones that are related to external dependencies that we can't/don't have in CI, or that other maintainers don't have (like various cloud providers, external services, etc.), this is _essential_.
  * labeling issues appropriately -- this helps everyone find relevant issues, and helps us see at a glance what a given issue is about.
* [answering questions](https://github.com/hvac/hvac/discussions) -- especially Q&A; a lot of these are troubleshooting, and if you are familiar with the project and with Vault and can answer these questions, your input is very valuable! If you aren't, then digging into these can be **very valuable to you** (and still us!) because you will become familiar with those things, and help someone else in the process.
* [documentation](https://github.com/hvac/hvac/issues?q=is%3Aopen+is%3Aissue+label%3Adocumentation) -- very important, and there's multiple ways to help with this
  * identifying errors or lack of clarity in docs (often this comes from Q&A and bug reports, especially ones that turn out to be misunderstandings, often due to... lacking or unclear docs!)
  * wordsmithing -- this includes: 
    * fixing typos
    * formatting -- the way text is formatted contributes _a lot_ to its readability (this post for example... could probably use some work 😅)
    * **grammar, phrasing, word choices** -- we are not currently a localised project, and this means that everyone who looks at our written words does so in English. There is a lot we can do to make English easier to read and understand for non-native speakers, which makes the project more accessible. **If you (especially as a non-native speaker) see some English that is difficult to understand, please open an issue even if you don't know how to fix it!**
  * adding examples -- unbelievably valuable, a code sample can be worth so much more than method definitions.
  * README, contributor guide, other meta documentation -- these are all gateways to using, contributing to, or otherwise interacting with the community and any improvements here are very welcome.
  * Defining documentation standards, style guides, conventions, etc. -- consistency throughout our communications is important, we don't have anything like this at the moment!
* [opening issues](https://github.com/hvac/hvac/issues) -- adding this as its own bullet because I want to say explicitly that reporting problems & suggestions is also contribution.
* [responding to polls](https://github.com/hvac/hvac/discussions/categories/polls) and requests for feedback -- Understanding how people use and interact with this software is something we have surprisingly little data about, and interaction with people who care about the project in any way directly helps determine the direction of the project and what decisions we make.
* code -- still valued and helpful
  * bugfixes -- especially ones that already have open issues because that's an indicator that it's affecting someone
  * features -- especially requested features like new authentication methods, or filling our existing secrets engines with missing methods or parameters
  * **tests** -- including improvements to existing tests, filling out missing coverage, coming up with ways to test tricky things
  * dependency and package management
  * CI/CD

## What does it mean to be a maintainer?
Honestly? It's _mostly_ ~all~ any of the above.

The rest is supporting fellow maintainers, and collaborating on decisions that determine the overall direction of the project.

I believe that we will find new maintainers organically via the contributors who are actively participating and are interested and willing.

See also, [The Open Source Contributor Funnel](https://mikemcquaid.com/the-open-source-contributor-funnel-why-people-dont-contribute-to-your-open-source-project/).

## I don't feel I'm qualified to do any of the above
That _might_ be true, and if it is it's ok, but I bet it's not!
If you can, I might encourage you to go directly to trying out some of the things. If it doesn't work out, that's perfectly fine!
If you're unsure and want to chat through it, reach out, I'd love to have that discussion.

## Suggestions welcome

Please feel free to reply to this topic with suggestions for more ways to get involved. I'm certain I missed some!

---

# How to get in touch about contributing

* [We have a Discord server for the project that you can join](https://discord.gg/HwM47nrGQU).
* Reply to this post

If you would like to reach out privately, at the moment a DM on Discord is probably the best way. 
We may look at other avenues. If Discord does not work for you, you may also reply to this post with a general suggestion, or with a specific way to get in touch with you.

**Comment by kaypeter87** [comment]
@briantist I can lend my services to help maintain this project. Let's chat separately.

**Comment by tot19** [comment]
I would be happy to jump in and assist where I can!

**Comment by klukashka** [comment]
Hello, @briantist!
I wrote to your email but you haven't answered yet. Here is my message:
My friend and me have been working on Python library for realtime configuration management, we needed a python client for Vault but we found out that hvac did not support async. So I began working on a new python client for Vault which would support both sync and async. The project is still being developed, it does not support some engines and authentication methods but basically it is usable.
I know I took the architecture and interface from hvac, so I am ready to mention it somehow in the license if it is necessary. I want to know if there is any sense in continuing the project.
I wanted to create a good library which would be used by someone.
I hope you appreciate my work. Please, tell me if it is so.
Hope, [Vaultx](https://github.com/klukashka/Vaultx) can be helpful anyhow and maybe I have ideas of how to improve hvac.

**Reply by briantist** [reply]
Hi @klukashka , I don't believe I've received any email from you but that's ok. You can also reach me on the hvac discord (link is at the bottom of the original post).

I think it's great to have an async version, there have been some requests for it in the past. I don't have much free time to look at it deeply but I'm sure it will be helpful to people. 

We can certainly chat about ideas to improve `hvac`, I would recommend opening new discussions here, or let's chat in the discord.

Thanks!
